Update the events, used with scripting, to use lower-case names and avoid using DOM events internally in the viewer + misc scripting-related tweaks

Snuffleupagus opened the pull request:
*Please refer to the individual commit messages for additional information.*

Also, ignoring white-space changes when viewing the larger patches should be helpful.

Snuffleupagus committed:
Tweak the new `mouseState` parameter, and its usage, in the viewer components and the `AnnotationLayer`

 - Actually remove the `isDown` property when destroying the scripting-instance.

 - Mark all `mouseState` usage as "private" in the various classes.

 - Ensure that the `AnnotationLayer` actually treats the parameter as properly *optional*, the same way that the viewer components do.

 - For now remove the `mouseState` parameter from the `PDFPageView` class, and keep it only on the `BaseViewer`, since it's questionable if all of the scripting-functionality will work all that well without e.g. a full `BaseViewer`.

 - Append the `mouseState` to the JSDoc for the `AnnotationElement` class, and just move its definition into the base-`AnnotationElement` class.

Snuffleupagus committed:
Tweak the `LinkAnnotationElement._bindJSAction` and `WidgetAnnotationElement.{_setEventListener, _setEventListeners}` methods

 - Update the `LinkAnnotationElement._bindJSAction` call-site to actually agree with the JSDocs, by passing in the `data`.

 - Prevent the links created by `LinkAnnotationElement._bindJSAction` from being displayed with empty hashes; compare with e.g. `LinkAnnotationElement. _bindNamedAction`.

 - The overall indentation-level in `WidgetAnnotationElement._setEventListener` can be reduced slightly by using early returns, which improves the overall readability of this method a bit. (We're also able to avoid unnecessary `in` usage here.)

 - The code can also be made *slightly* more efficient overall, by moving the `this.data.actions` check into `WidgetAnnotationElement._setEventListeners` instead. This way we can avoid useless `this._setEventListener`-calls when there are no actions present.

Snuffleupagus committed:
Update the events, used with scripting, to use lower-case names and avoid using DOM events internally in the viewer

For DOM events all event names are lower-case, and the newly added PDF.js scripting-events thus "stick out" quite a bit. Even more so, considering that our internal `eventBus`-events follow the same naming convention.
Hence this patch, which changes the "updateFromSandbox"/"dispatchEventInSandbox" events to be lower-case instead.

Furthermore, using DOM events for communication *within* the PDF.js code itself (i.e. between code in `web/app.js` and `src/display/annotation_layer.js/`) feels *really* out of place.
That's exactly the reason that we have the `EventBus` abstraction, since it allowed us to remove prior use of DOM events, and this patch thus re-factors the code to make use of the `EventBus` instead for scripting-related events.
Obviously for events targeting a *specific element* using DOM events is still fine, but the "updatefromsandbox"/"dispatcheventinsandbox" ones should be using the `EventBus` internally.

*Drive-by change:* Use the `BaseViewer.currentScaleValue` setter unconditionally in `PDFViewerApplication._initializeJavaScript`, since it accepts either a string or a number.

Snuffleupagus committed:
Remove the `ENABLE_SCRIPTING` build-target, since it's not necessary

There's no really compelling reason, as far as I can tell, to introduce the `ENABLE_SCRIPTING` build-target, instead of simply re-using the existing `TESTING` build-target for the new `gulp integrationtest` task.

In general there should be no problem with just always enable scripting in TESTING-builds, and if I were to *guess* the reason that this didn't seem to work was most likely because the Preferences ended up over-writing the `AppOptions`.
As it turns out the GENERIC-viewer has already has built-in support for disabling of Preferences, via the `AppOptions`, and this can be utilized in TESTING-builds as well to ensure that whatever `AppOptions` are set they're always respected.

Snuffleupagus commented:
/botio unittest

calixteman commented on src/display/annotation_layer.js:
```diff
@@ -544,40 +544,42 @@ class WidgetAnnotationElement extends AnnotationElement {
+    if (baseName.includes("mouse")) {
+      // Mouse events
+      element.addEventListener(baseName, event => {
+        this.linkService.eventBus?.dispatch("dispatcheventinsandbox", {
```
I think we should have at least a warning in the console if eventBus is null.
Or if we expect that eventBus is neither null nor undefined, then you can remove the `?` and we'll get an exception.

Snuffleupagus replied:
That's not really a problem though, since the `AnnotationLayer` *must* be initialized with either a `PDFLinkService` or a `SimpleLinkService` instance (the latter having only dummy-methods).
Note how the existing code simply assumes that *either* is available, and the optional chaining added here (and elsewhere) is simply intended to prevent errors in the `SimpleLinkService`-case.

Snuffleupagus commented:
/botio integrationtest

calixteman commented on src/display/annotation_layer.js:
```diff
@@ -717,39 +718,37 @@ class TextWidgetAnnotationElement extends WidgetAnnotationElement {
+          const _blurListener = blurListener;
+          blurListener = null;
+          element.addEventListener("blur", event => {
+            if (this._mouseState?.isDown) {
```
`_mouseState` shoudn't be null here else the behaviour won't be correct, so you should report something (a warning or throw an exception).

Snuffleupagus replied:
We're not currently warning/throwing for *any* of the parameters used with the `AnnotationLayer`, which assuming we even care enough to fix this (which is somewhat doubtful to me), should then probably be fixed all at once rather than for just one parameter.

Please note also that it's perfectly fine for the `mouseState` parameter to be optional in *some* cases:
 - The `AnnotationLayer` is also utilized with the viewer-components, which can be used build a simplified viewer without all the bells and whistles of the full default viewer, and in that case we *really* shouldn't require all of the scripting-related options to be provided by the user.
 - Furthermore, the *existing* viewer JSDocs specifically listed the `mouseState`-parameter as optional in e.g. the `BaseViewer` and consequently it should be optional *everywhere* including the `AnnotationLayer`.

However, I've changed things slightly to ensure that `this._mouseState` is always defined with the expected format, such that we don't need to use inline checks in (potentially) multiple places.

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented:
/botio test

Snuffleupagus committed:
Don't dispatch a "doc/Open" event in the sandbox when creating it failed

There's really no point, as far as I can tell, to attempt to dispatch an event in a non-existent sandbox. Generally speaking, even trying to do this *could* possibly even lead to errors in some cases.

Furthermore, utilize optional chaining to simplify some `dispatchEventInSandbox` calls throughout the viewer.

Finally, replace superfluous `return` statements with `break` in the switch-statement in the `updateFromSandbox` event-handler.

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus committed:
Remove the arbitrary timeout in the "must check that first text field has focus" integration-test (PR 12702 follow-up)

It seems that the timeout is way too short in practice, since this new integration-test failed *intermittently* already in PR 12702 (which is where the test was added).

The ideal solution here would be to simply await an event, dispatched by the viewer, however that unfortunately doesn't appear to be supported by Puppeteer.
Instead, the solution implemented here is to add a new method in `PDFViewerApplication` which Puppeteer can query to check if the scripting/sandbox has been fully initialized.

calixteman commented:
The failure in integration test is because the open action is executed after trying to get the focused element.
I'm figuring out a way to fix it.

Snuffleupagus commented:
/botio integrationtest

calixteman commented:
@Snuffleupagus I think this patch fix the issue: https://github.com/mozilla/pdf.js/pull/12755

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented:
/botio test

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
Thank you for improving the overall consistency here!
